src/Storages/TableZnodeInfo.cpp

```diff
@@ -0,0 +1,135 @@
+    {
+        auto database = DatabaseCatalog::instance().getDatabase(table_id.database_name);
+        info.shard = getReplicatedDatabaseShardName(database);
+        info.replica = getReplicatedDatabaseReplicaName(database);
+    }
+    if (!allow_uuid_macro)
+        info.table_id.uuid = UUIDHelpers::Nil;
+    res.full_path = context->getMacros()->expand(res.full_path, info);
+    bool expanded_uuid_in_path = info.expanded_uuid;
+
+    info.level = 0;
+    info.table_id.uuid = UUIDHelpers::Nil;
+    res.replica_name = context->getMacros()->expand(res.replica_name, info);
+
+    /// We do not allow renaming table with these macros in metadata, because zookeeper_path will be broken after RENAME TABLE.
+    /// NOTE: it may happen if table was created by older version of ClickHouse (< 20.10) and macros was not unfolded on table creation
+    /// or if one of these macros is recursively expanded from some other macro.
+    /// Also do not allow to move table from Atomic to Ordinary database if there's {uuid} macro
+    if (info.expanded_database || info.expanded_table)
+        res.renaming_restrictions = RenamingRestrictions::DO_NOT_ALLOW;
+    else if (info.expanded_uuid)
+        res.renaming_restrictions = RenamingRestrictions::ALLOW_PRESERVING_UUID;
+
+    res.zookeeper_name = zkutil::extractZooKeeperName(res.full_path);
+    res.path = zkutil::extractZooKeeperPath(res.full_path, /* check_starts_with_slash */ mode <= LoadingStrictnessLevel::CREATE, getLogger(table_id.getNameForLogs()));
+    res.path_prefix_for_drop = res.path;
+
+    if (expanded_uuid_in_path)
+    {
+        /// When dropping table with znode path "/foo/{uuid}/bar/baz", delete not only
+        /// "/foo/{uuid}/bar/baz" but also "/foo/{uuid}/bar" and "/foo/{uuid}" if they became empty.
+        ///
+        /// (We find the uuid substring by searching instead of keeping track of it when expanding
+        ///  the macro. So in principle we may find a uuid substring that wasn't expanded from a
+        ///  macro. This should be ok because we're searching for the *last* occurrence, so we'll get
+        ///  a prefix at least as long as the correct one, so we won't delete znodes outside the
+        ///  {uuid} path component. This sounds sketchy, but propagating string indices through macro
+        ///  expansion passes is sketchy too (error-prone and more complex), and on balance this seems
+        ///  better.)
+        String uuid_str = toString(table_id.uuid);
+        size_t i = res.path.rfind(uuid_str);
+        if (i == String::npos)
+            /// Possible if the macro is in the "<auxiliary_zookeeper_name>:/" prefix, but we probably
+            /// don't want to allow that.
+            throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't find uuid in zookeeper path after expanding {{uuid}} macro: {} (uuid {})", res.path, uuid_str);
+        i += uuid_str.size();
+        /// In case the path is "/foo/pika{uuid}chu/bar" (or "/foo/{uuid}{replica}/bar").
```

Comment:
Should a `/foo/pikachu{uuid}/` prefix also be filtered out?

Reply:
`i` will end up at the end of the path component containing the uuid (right?), and that's what we want (right?). `{A}/{B}{uuid}{C}/{D}` -> `{A}/{B}{uuid}{C}`, for any (possibly empty) strings A, B, C, D (A and B not containing slashes).
 * `/foo/pikachu{uuid}/[...]` -> `/foo/pikachu{uuid}`
 * `/foo/pika{uuid}pika/[...]` -> `/foo/pika{uuid}pika`